File: chapter-07-automation/python/02-delta-queries/delta_queries.py
#!/usr/bin/env python3
"""
delta_queries.py
Ejemplo de Delta Queries con Microsoft Graph para sincronizacion incremental.
Referencia: Capitulo 7 - Automatizacion y Flujos

Uso:
    export CLIENT_ID="your-client-id"
    export CLIENT_SECRET="your-client-secret"
    export TENANT_ID="your-tenant-id"
    python delta_queries.py
"""

import os
import sys
from datetime import datetime
from typing import Optional, Dict, List, Any

import requests


class DeltaQueryManager:
    """Gestiona consultas delta para sincronizacion incremental."""

    def __init__(self, access_token: str):
        self.access_token = access_token
        self.base_url = "https://graph.microsoft.com/v1.0"
        self.delta_tokens: Dict[str, str] = {}

    def _get_headers(self) -> Dict[str, str]:
        return {
            "Authorization": f"Bearer {self.access_token}",
            "Accept": "application/json"
        }

    def get_initial_delta(self, drive_id: str, folder_id: str = "root") -> List[Dict]:
        """Realiza consulta delta inicial y almacena token."""
        print(f"🔄 Ejecutando consulta delta inicial en drive: {drive_id}")

        url = f"{self.base_url}/drives/{drive_id}/items/{folder_id}/delta"
        response = requests.get(url, headers=self._get_headers())
        response.raise_for_status()

        data = response.json()
        items = data.get("value", [])
        print(f"   Items obtenidos: {len(items)}")

        # Guardar token delta
        delta_link = data.get("@odata.deltaLink", "")
        token = self._extract_delta_token(delta_link)
        if token:
            key = f"{drive_id}_{folder_id}"
            self.delta_tokens[key] = token
            print(f"   Token delta guardado: {token[:50]}...")

        return items

    def get_delta_changes(self, drive_id: str, folder_id: str = "root") -> List[Dict]:
        """Sincroniza cambios usando token delta almacenado."""
        key = f"{drive_id}_{folder_id}"

        if key not in self.delta_tokens:
            print("⚠️  No existe token delta. Ejecutando consulta inicial...")
            return self.get_initial_delta(drive_id, folder_id)

        token = self.delta_tokens[key]
        print(f"🔄 Sincronizando cambios con token delta...")

        url = f"{self.base_url}/drives/{drive_id}/items/{folder_id}/delta"
        params = {"token": token}

        response = requests.get(url, headers=self._get_headers(), params=params)
        response.raise_for_status()

        data = response.json()
        items = data.get("value", [])
        print(f"   Cambios detectados: {len(items)}")

        for item in items:
            if "deleted" in item:
                print(f"   🗑️  Eliminado: {item.get('name', 'unknown')}")
            else:
                print(f"   ✏️  Modificado: {item.get('name', 'unknown')}")

        # Actualizar token delta
        new_delta_link = data.get("@odata.deltaLink", "")
        new_token = self._extract_delta_token(new_delta_link)
        if new_token:
            self.delta_tokens[key] = new_token
            print(f"   Token delta actualizado")

        # Verificar si hay mas paginas
        next_link = data.get("@odata.nextLink")
        if next_link:
            print(f"   📄 Hay mas paginas. Usa: {next_link}")

        return items

    def get_list_delta(self, site_id: str, list_id: str) -> List[Dict]:
        """Consulta delta para items de una lista de SharePoint."""
        print(f"🔄 Consulta delta para lista: {list_id}")

        url = f"{self.base_url}/sites/{site_id}/lists/{list_id}/items/delta"
        response = requests.get(url, headers=self._get_headers())
        response.raise_for_status()

        data = response.json()
        items = data.get("value", [])

        # Guardar token
        delta_link = data.get("@odata.deltaLink", "")
        token = self._extract_delta_token(delta_link)
        if token:
            self.delta_tokens[f"list_{site_id}_{list_id}"] = token

        return items

    def _extract_delta_token(self, delta_link: str) -> Optional[str]:
        """Extrae el token delta de la URL."""
        if not delta_link:
            return None

        from urllib.parse import urlparse, parse_qs
        parsed = urlparse(delta_link)
        params = parse_qs(parsed.query)

        return params.get("token", [None])[0] or params.get("deltaToken", [None])[0]

    def show_stored_tokens(self):
        """Muestra los tokens delta almacenados."""
        print("\n📋 Tokens Delta Almacenados:")
        for key, token in self.delta_tokens.items():
            print(f"   {key}: {token[:50]}...")


def main():
    print("╔══════════════════════════════════════════════════════════════╗")
    print("║     EJEMPLO: DELTA QUERIES CON MICROSOFT GRAPH              ║")
    print("║     Sincronizacion Incremental de Documentos                 ║")
    print("╚══════════════════════════════════════════════════════════════╝\n")

    from laexperiencia_sharepoint import get_access_token

    access_token = get_access_token()
    manager = DeltaQueryManager(access_token)

    hostname = os.getenv("SHAREPOINT_HOSTNAME", "olddogsoft1.sharepoint.com")
    site_path = os.getenv("SHAREPOINT_SITE_PATH", "book-test")

    try:
        # Resolver el sitio por path y su primera biblioteca.
        site_url = (
            f"{manager.base_url}/sites/"
            f"{requests.utils.quote(hostname, safe='')}:/sites/"
            f"{requests.utils.quote(site_path, safe='')}"
        )
        site = requests.get(site_url, headers=manager._get_headers())
        site.raise_for_status()
        site_id = site.json()["id"]

        drives = requests.get(f"{manager.base_url}/sites/{site_id}/drives",
                              headers=manager._get_headers())
        drives.raise_for_status()
        drive_id = drives.json()["value"][0]["id"]
        print(f"Biblioteca objetivo: {drives.json()['value'][0]['name']}\n")

        # Primera ejecucion: consulta inicial
        print("--- Consulta Delta Inicial ---")
        manager.get_initial_delta(drive_id)

        # Segunda ejecucion: solo cambios
        print("\n--- Sincronizacion de Cambios ---")
        manager.get_delta_changes(drive_id)

        manager.show_stored_tokens()
        print("\n✅ Ejemplo completado")

    except requests.HTTPError as e:
        print(f"❌ Error HTTP: {e.response.status_code} - {e.response.text}")
    except Exception as e:
        print(f"❌ Error: {e}")


if __name__ == "__main__":
    main()

File: chapter-07-automation/javascript/02-delta-queries/deltaQueries.js
/**
 * deltaQueries.js
 * Example of Delta Queries with Microsoft Graph for incremental sync.
 * Reference: Chapter 7 - Automation and Flows
 *
 * Usage:
 *   node deltaQueries.js
 */

const axios = require('axios');
const { getAccessToken } = require('la-experiencia-sharepoint-code/graphAuth');

const GRAPH_BASE = 'https://graph.microsoft.com/v1.0';

class DeltaQueryManager {
    constructor(accessToken) {
        this.accessToken = accessToken;
        this.baseUrl = 'https://graph.microsoft.com/v1.0';
        this.deltaTokens = new Map();
    }

    _getHeaders() {
        return {
            'Authorization': `Bearer ${this.accessToken}`,
            'Accept': 'application/json'
        };
    }

    /**
     * Realiza consulta delta inicial y almacena token.
     */
    async getInitialDelta(driveId, folderId = 'root') {
        console.log(`🔄 Ejecutando consulta delta inicial en drive: ${driveId}`);

        const url = `${this.baseUrl}/drives/${driveId}/items/${folderId}/delta`;
        const response = await axios.get(url, { headers: this._getHeaders() });

        const items = response.data.value || [];
        console.log(`   Items obtenidos: ${items.length}`);

        // Guardar token delta
        const deltaLink = response.data['@odata.deltaLink'] || '';
        const token = this._extractDeltaToken(deltaLink);
        if (token) {
            const key = `${driveId}_${folderId}`;
            this.deltaTokens.set(key, token);
            console.log(`   Token delta guardado: ${token.substring(0, 50)}...`);
        }

        return items;
    }

    /**
     * Sincroniza cambios usando token delta almacenado.
     */
    async getDeltaChanges(driveId, folderId = 'root') {
        const key = `${driveId}_${folderId}`;

        if (!this.deltaTokens.has(key)) {
            console.log('⚠️  No existe token delta. Ejecutando consulta inicial...');
            return await this.getInitialDelta(driveId, folderId);
        }

        const token = this.deltaTokens.get(key);
        console.log('🔄 Sincronizando cambios con token delta...');

        const url = `${this.baseUrl}/drives/${driveId}/items/${folderId}/delta`;
        const response = await axios.get(url, {
            headers: this._getHeaders(),
            params: { token }
        });

        const items = response.data.value || [];
        console.log(`   Cambios detectados: ${items.length}`);

        items.forEach(item => {
            if (item.deleted) {
                console.log(`   🗑️  Deleted: ${item.name || 'unknown'}`);
            } else {
                console.log(`   ✏️  Modificado: ${item.name || 'unknown'}`);
            }
        });

        // Update delta token
        const newDeltaLink = response.data['@odata.deltaLink'] || '';
        const newToken = this._extractDeltaToken(newDeltaLink);
        if (newToken) {
            this.deltaTokens.set(key, newToken);
            console.log('   Token delta actualizado');
        }

        // Verificar paginacion
        const nextLink = response.data['@odata.nextLink'];
        if (nextLink) {
            console.log(`   📄 Hay mas paginas: ${nextLink}`);
        }

        return items;
    }

    /**
     * Consulta delta para items de una lista de SharePoint.
     */
    async getListDelta(siteId, listId) {
        console.log(`🔄 Consulta delta para lista: ${listId}`);

        const url = `${this.baseUrl}/sites/${siteId}/lists/${listId}/items/delta`;
        const response = await axios.get(url, { headers: this._getHeaders() });

        const items = response.data.value || [];

        const deltaLink = response.data['@odata.deltaLink'] || '';
        const token = this._extractDeltaToken(deltaLink);
        if (token) {
            this.deltaTokens.set(`list_${siteId}_${listId}`, token);
        }

        return items;
    }

    /**
     * Extrae el token delta de la URL.
     */
    _extractDeltaToken(deltaLink) {
        if (!deltaLink) return null;

        const url = new URL(deltaLink);
        return url.searchParams.get('token') || url.searchParams.get('deltaToken');
    }

    /**
     * Shows the stored delta tokens.
     */
    showStoredTokens() {
        console.log('\n📋 Tokens Delta Almacenados:');
        this.deltaTokens.forEach((token, key) => {
            console.log(`   ${key}: ${token.substring(0, 50)}...`);
        });
    }
}

async function main() {
    console.log('╔══════════════════════════════════════════════════════════════╗');
    console.log('║     EJEMPLO: DELTA QUERIES CON MICROSOFT GRAPH             ║');
    console.log('║     Sincronizacion Incremental de Documentos                 ║');
    console.log('╚══════════════════════════════════════════════════════════════╝\n');

    const accessToken = await getAccessToken();
    const manager = new DeltaQueryManager(accessToken);

    const hostname = process.env.SHAREPOINT_HOSTNAME || 'olddogsoft1.sharepoint.com';
    const sitePath = process.env.SHAREPOINT_SITE_PATH || 'book-test';

    try {
        // Resolver el sitio por path y su primera biblioteca.
        const siteUrl = `${GRAPH_BASE}/sites/${encodeURIComponent(hostname)}:/sites/${encodeURIComponent(sitePath)}`;
        const siteResp = await axios.get(siteUrl, { headers: manager._getHeaders() });
        const siteId = siteResp.data.id;
        const drivesResp = await axios.get(`${GRAPH_BASE}/sites/${siteId}/drives`, { headers: manager._getHeaders() });
        const firstDrive = drivesResp.data.value[0];
        console.log(`Biblioteca objetivo: ${firstDrive.name}\n`);

        // Consulta inicial
        console.log('--- Consulta Delta Inicial ---');
        await manager.getInitialDelta(firstDrive.id);

        // Sincronizacion de cambios
        console.log('\n--- Sincronizacion de Cambios ---');
        await manager.getDeltaChanges(firstDrive.id);

        manager.showStoredTokens();
        console.log('\n✅ Example completed');

    } catch (error) {
        console.error(`❌ Error: ${error.message}`);
        if (error.response) {
            console.error(`   Status: ${error.response.status}`);
            console.error(`   Data: ${JSON.stringify(error.response.data, null, 2)}`);
        }
    }
}

module.exports = { DeltaQueryManager };

if (require.main === module) {
    main();
}
